luda-editor/client/src/pages/graphic_asset_manage_page/cg_viewer/mod.rs

```diff
@@ -0,0 +1,269 @@
+mod part_picker;
+
+use self::part_picker::PartPicker;
+use crate::{color, components::cg_render::CgRender, storage::get_project_cg_thumbnail_image_url};
+use namui::prelude::*;
+use namui_prebuilt::{button, simple_rect, table::hooks::*, typography};
+use rpc::data::{ScreenCg, ScreenCgPart};
+use std::ops::Deref;
+
+const MODAL_MAX_WH: Wh<Px> = Wh {
+    width: px(1280.0),
+    height: px(720.0),
+};
+const MODAL_MIN_MARGIN: Px = px(16.0);
+const TITLE_BAR_HEIGHT: Px = px(48.0);
+const PICKER_WIDTH: Px = px(512.0);
+
+#[component]
+pub struct CgViewer<'a> {
+    pub wh: Wh<Px>,
+    pub cg_file: &'a rpc::data::CgFile,
+    pub project_id: Uuid,
+    pub on_close: &'a dyn Fn(),
+}
+
+impl Component for CgViewer<'_> {
+    fn render(self, ctx: &RenderCtx) -> RenderDone {
+        let Self {
+            wh,
+            cg_file,
+            project_id,
+            on_close,
+        } = self;
+
+        let wh = ctx.track_eq(&wh);
+        let modal_rect = ctx.memo(|| {
+            let modal_wh = Wh {
+                width: (wh.width - MODAL_MIN_MARGIN * 2.0).min(MODAL_MAX_WH.width),
+                height: (wh.height - MODAL_MIN_MARGIN * 2.0).min(MODAL_MAX_WH.height),
+            };
+            let modal_xy = Xy {
+                x: (wh.width - modal_wh.width) / 2.0,
+                y: (wh.height - modal_wh.height) / 2.0,
+            };
+            Rect::from_xy_wh(modal_xy, modal_wh)
+        });
+        let (screen_cg, set_screen_cg) = ctx.state(|| ScreenCg::new(cg_file));
```

Comment:
Is it right that cg_file is only for initialization?

Reply:
No, cg_file will also be used to render preview and to modifying screen_cg on preview.

Reply:
But your code initialize screen_cg with cg_file in ctx.state. Did you tracked cg_file's changes and update screen_cg?

Reply:
No I didn't tracked that.

Reply:
Fixed with 5215372

Reply:
This code seems very weird. Think about this like react.
You want to change screen cg without editing file directly until editing finished, and update file when user completes editing, right?
Then it should be ignored the cg_file's changes during editing.

Reply:
No, We don't change cg_file with this **preview**.
`cg_file` change here means "User wants to see other cg_file's preview"

Reply:
So, it's only for initialization of screen_cg, right?

Reply:
Yes

Reply:
> Is it right that cg_file is only for initialization?

That's what I'm saying at the first time.

Put the comment or change the name to describe this is only for initialization so it won't change initialized cg_viewer if user change cg_file.

Reply:
I put comment #324432e

Reply:
Put comment to props parameter, not the state.
It's for viewer component user, not for viewer developer.

Reply:
moved 5265b58